#10 버스터미널 조회 기능 추가

0hun opened the pull request:
버스터미널 조회 기능 추가
기존 소스 수정

0hun committed:
#10 버스터미널 조회 기능 추가
기존 소스 수정

0hun committed:
#10 버스 터미널 조회 수정

return ResponseEntity에 body에 busTerminal 넣도록 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/BusTerminalController.java:
```diff
@@ -0,0 +1,43 @@
+   * @param busTerminalDto 조회할 터미널 이름, 지역 이름
+   * @return
+   */
+  @GetMapping("/terminal")
```
여기와 아래의 들여쓰기가 맞지 않네요~ check style은 커밋하기전에 계속 돌려주세요~
checkstyle 검사를 `commit hook`에 등록해서 쓰셔도 좋습니다~

f-lab-dev replied:
REST한 URL구조를 만들면 좋을 것 같습니다~ 이 자료를 공부해보시면 좋을 것 같아요

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

0hun replied:
넵 알겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/BusTerminalController.java:
```diff
@@ -0,0 +1,43 @@
+   * @return
+   */
+  @GetMapping("/terminal")
+    public ResponseEntity<BusTerminal> getBusTerminal(@RequestBody BusTerminalDto busTerminalDto) {
```
GET에서는 일반적으로 body를 사용하지 않습니다~

0hun replied:
그럼 GET method에서 PathVariable로 2개의 매개변수로 받는식으로 수정하면 될까요

0hun replied:
넵 수정하겠습니다.

f-lab-dev replied:
제가 드린 URL 설계규칙을 보시면서 어떤걸 path로 받을지, 어떤걸 query로 받을지 고민해보시는게 좋겠습니다~

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/BusTerminalController.java:
```diff
@@ -0,0 +1,43 @@
+      BusTerminal busTerminal = busTerminalService.findByNameAndRegion(name, region);
+
+      if (busTerminal == null) {
+        return RESPONSE_ENTITY_NO_CONTENT;
```
조회한 리소스가 없으면 클라이언트가 없는 버스정류장을 조회한거니 404를 리턴해줘야하지 않을까요?

0hun replied:
그럼 혹시 앞서 로그인 메서드에서 유저가 없을시 RESPONSE_ENTITY_NO_CONTENT(204) code를 리턴했는데, 이 경우에도 
404리턴이 맞는건가요?

f-lab-dev replied:
네 NO_CONTENT는 다른 의미입니다~

f-lab-dev commented on src/main/java/com/younghun/hibusgo/domain/BusTerminal.java:
```diff
@@ -0,0 +1,29 @@
+import lombok.ToString;
+
+@Getter
+@Setter
```
이 클래스는 setter가 꼭 필요한가요? `java immutable`에 대해 알아보시면 좋을 것 같습니다~

0hun replied:
넵 알겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/domain/BusTerminal.java:
```diff
@@ -0,0 +1,29 @@
+
+    private String region;
+
+    // 상태
```
`status`라는 이름에서 상태라는 뜻을 유추할 수 있습니다. 주석은 용도를 명확하게 알 수 있도록 적어주세요~

0hun replied:
넵 수정하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/BusTerminalService.java:
```diff
@@ -0,0 +1,37 @@
+
+    private final BusTerminalMapper terminalMapper;
+
+    public BusTerminal findById(int id) {
```
이런 데이터들은 최신성이 보장될 필요가 없으므로 캐싱을 적용하면 좋을 것 같습니다~
별도의 이슈를 만들어서 작업해보시면 좋을 것 같네요

0hun replied:
redis에 저장하는걸 말하는건가요?

f-lab-dev replied:
캐싱은 레디스에 저장할수도 있고 로컬메모리에 저장할 수도 있습니다. 어떤걸 사용할지는 고민해보시는게 좋겠네요~
`spring cache`를 사용하시면 좋을 것 같고요

0hun replied:
넵 알겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/validator/PasswordValidator.java:
```diff
@@ -15,6 +20,7 @@ public boolean supports(Class<?> clazz) {
   public void validate(Object target, Errors errors) {
     PasswordDto passwordDto = (PasswordDto) target;
     if (!passwordDto.getNewPassword().equals(passwordDto.getNewPasswordConfirm())) {
+      log.error("입력한 새 패스워드가 일치하지 않습니다.");
```
이 로그는 개발자가 보는 것인데 누가 패스워드를 틀렸다는 로그를 모두 찍어줄 필요가 있을까요?
보안상 꼭 찍어줘야하는 로그라면 누가 실패했는지 정보를 같이 찍어줘야 해당 유저의 행동을 추적해볼 수 있지 않을까요?

0hun replied:
개발자가 보려고 만든 로그 목적으로 사용한것이니 꼭 필요한 상황에서만 써야되는건가요?

f-lab-dev replied:
대용량 트래픽이 들어온다고 가정하면 엄청난 양의 로그가 쌓이게 될텐데 이런 경우를 위해서라도 꼭 필요한 로그만 찍어주는게 좋습니다.
그러지 않으면 로그 저장비용도 엄청 나올테고 에러가 발생했을때도 개발자가 파악할 수 없기때문에 로그의 의미가 없어집니다

f-lab-dev commented on src/main/resources/mapper/busTerminal.xml:
```diff
@@ -0,0 +1,46 @@
+
+    <select id="findById" resultMap="accountResultMap">
+        SELECT b.id
+        , b.name
```
쉼표는 라인의 시작이 아니라 마지막에 찍어주면 더 깔끔하지 않을까요?

0hun replied:
넵 수정하겠습니다.

0hun replied:
앞선 account.xml 내용에선 쉼표 위치에 대해 언급이 없으셔서 맞는 스타일로 생각하고 있었습니다.

f-lab-dev commented on src/main/resources/mapper/busTerminal.xml:
```diff
@@ -0,0 +1,46 @@
+        , b.status
+        FROM busTerminal b
+        WHERE b.id = #{id}
+        AND status != 'DELETE'
```
`findById` 라는 쿼리는 재사용성이 꽤 높아보입니다. 삭제된 계정도 조회할 가능성도 있을 것 같은데 status가 `DELETE`라는건 자바쪽에서 검사해주면 어떨까요?
만약 이 쿼리에서 리턴결과가 없으면 DELETE라서 결과가 없는건지 유저가 없는건지 헷갈릴 수도 있고요~

0hun replied:
넵 수정하겠습니다.

0hun replied:
그럼 앞선 계정을 조회하는 문제에서도 status를 자바 로직에서 체크하도록 수정해야할까요?

f-lab-dev replied:
네 물론 상황에 따라 다르기때문에 계속 고민해보시는게 좋을 것 같네요

f-lab-dev commented on src/main/resources/mapper/busTerminal.xml:
```diff
@@ -0,0 +1,46 @@
+        <if test="region !=null and region !=''">
+            AND  region LIKE '%' || #{region} || '%'
+        </if>
+        AND status != 'DELETE'
```
1. 이 쿼리의 결과가 아주 많아질 가능성은 없나요? 
2. LIKE 연산을 할 대상이 너무 많으면 연산이 너무 오래걸리지 않을까요?
3. 검색조건에 적절한 인덱스가 걸려있나요?

0hun replied:
1.지역정보만 들어가기때문에 많아질 가능성은 적을것 같습니다.
2.연산 대상이 많으면 연산이 오래걸릴꺼같습니다.
3.인덱스는 걸려있습니다.

그럼 like 연산은 사용하지 않는편이 좋은가요?

0hun replied:
https://jojoldu.tistory.com/243 [mysql 인덱스 정리 및 팁]

0hun replied:
인덱스가 걸려있어도 full scan을 하게되면 느려지게 될 가능성이 있다는 의미 인가요?

f-lab-dev replied:
네 인덱스를 통해 검색결과의 범위를 얼마나 줄이냐를 고려하셔야합니다~

0hun committed:
#10 버스터미널 조회 기능 추가 리뷰 반영

getBusTerminal pathvariable로 매개 변수 받도록 수정
REST한 URL구조로 변경
busTerminal null 또는 Status.DELETED일 경우 404 return
불필요하거나 명확하지 않은 log 삭제
xml 쉼표 깔끔하게 라인의 마지막으로 위치 변경
status DELETE는 삭제된 데이터도 조회 할 수로 도 있으므로 자바 로직에서 체크

0hun committed:
#10 버스터미널 조회 기능 추가 리뷰 반영

체크 스타일 적용 및 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/LoginUtil.java:
```diff
@@ -63,7 +63,7 @@ public boolean isLoginAccount() {
   @Override
   public String getLoginAccountId() {
-    Optional<String> accountId = (Optional<String>) session.getAttribute(ACCOUNT_MEMBER_ID);
+    Optional<String> accountId = Optional.ofNullable((String) session.getAttribute(ACCOUNT_MEMBER_ID));
```
`session.getAttribute(ACCOUNT_MEMBER_ID)`가 null일 수 있으므로 일단 Optional로 만들어준 후 `map` 메소드를 이용해 형변환해주는게 어떨까요?

0hun replied:
넵 수정하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/ResponseConstants.java:
```diff
@@ -10,8 +10,8 @@
+  public static final ResponseEntity<?> RESPONSE_ENTITY_NO_CONTENT = ResponseEntity.noContent().build();
+  public static final ResponseEntity<?> RESPONSE_ENTITY_CREATED = ResponseEntity.status(HttpStatus.CREATED).build();
+  public static final ResponseEntity<?> RESPONSE_ENTITY_OK = ResponseEntity.ok().build();
+  public static final ResponseEntity<?> RESPONSE_NOT_FOUND = ResponseEntity.notFound().build();
```
이 상수들은 모두 body 타입이 고정적인데 모든 타입을 담을 수 있는 제네릭 타입보다는 아무것도 없다는 제네릭타입으로 선언해주시는게 좋아보입니다

0hun replied:
넵 수정하겠습니다.

0hun committed:
#10 버스터미널 조회 기능 추가 리뷰 반영

session.getAttribute(ACCOUNT_MEMBER_ID)가 null일 수 있으므로Optional로 만들어준 후 map 메소드를 이용해 형변환

ResponseConstants 상수가 리턴이 고정된 타입이기 떄문에 아무것도 없는 제네릭 타입으로 변경

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/LoginUtil.java:
```diff
@@ -63,7 +63,8 @@ public boolean isLoginAccount() {
-    return accountId.orElse(null);
+    return Optional.ofNullable(session.getAttribute(ACCOUNT_MEMBER_ID))
+        .map(String::valueOf)
+        .orElse(null);
```
이런 경우는 null을 리턴해주기보다는 `Optional` 타입 통째로 리턴하셔서 결과가 없을수도 있음을 메소드 사용자에게 알려주는 것이 좋습니다~

0hun replied:
넵 알겠습니다.

0hun replied:
erd 리뷰 부탁드립니다.읽기 전용 url(비밀번호 5ez653) https://aquerytool.com:443/aquerymain/index/?rurl=70475dc0-1cb5-4573-a864-93e68f4ddcaa

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/ResponseConstants.java:
```diff
@@ -10,8 +10,8 @@
-  public static final ResponseEntity<?> RESPONSE_ENTITY_CREATED = ResponseEntity.status(HttpStatus.CREATED).build();
-  public static final ResponseEntity<?> RESPONSE_ENTITY_OK = ResponseEntity.ok().build();
-  public static final ResponseEntity<?> RESPONSE_NOT_FOUND = ResponseEntity.notFound().build();
+  public static final ResponseEntity RESPONSE_ENTITY_NO_CONTENT = ResponseEntity.noContent().build();
```
제네릭을 없애라는 뜻이 아니라 아무것도 없다는 것도 제네릭으로 명시할 수 있습니다. 어떻게 하면 좋을까요?

0hun committed:
#10 버스티머닐 조회 기능 추가 리뷰 반영

터미널 조회시 status 체크 서비스 로직에서 하도록 변경
로그인 사용자 정보 가져올때 optional return
ResponseConstants 제네릭 void 타입으로 변경

f-lab-dev commented on src/test/java/com/younghun/hibusgo/controller/AccountControllerTest.java:
```diff
@@ -87,7 +87,7 @@ void addAccount_with_correct_input() throws Exception {
 
         Account account = accountDto.toEntity();
-        account.setStatus(Account.Status.DEFAULT);
+//        account.setStatus(Account.Status.DEFAULT);
```
주석은 커밋하지 않는게 좋습니다~ 소스가 머지된다면 최종 소스에 불필요한 소스가 포함되기 때문입니다

0hun replied:
넵 알겠습니다

f-lab-dev reviewed (approved):
주석부분 리뷰드린것만 수정하시면 머지하셔도 됩니다~

0hun committed:
#10 리뷰 반영

주석 삭제
